bigquery/v2/query/query.go

```diff
@@ -0,0 +1,294 @@
+// cancel the wait. If the query completes successfully, Wait returns nil.
+// Otherwise, it returns the error that caused the query to fail.
+//
+// Wait is a convenience wrapper around Done and Err.
+func (q *Query) Wait(ctx context.Context) error {
+	select {
+	case <-q.Done():
+		return q.Err()
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Done returns a channel that is closed when the query has completed.
+// It can be used in a select statement to perform non-blocking waits.
+//
+// Example:
+//
+//		select {
+//		case <-q.Done():
+//			if err := q.Err(); err != nil {
+//				// Handle error.
+//			}
+//			// Query is complete.
+//	 case <-time.After(30*time.Second):
+//		    // Timeout logic
+//		default:
+//			// Query is still running.
+//		}
+func (q *Query) Done(opts ...gax.CallOption) <-chan struct{} {
+	return q.ready
+}
+
+// Err returns the final error state of the query. It is only valid to call Err
+// after the channel returned by Done has been closed. If the query completed
+// successfully, Err returns nil.
+func (q *Query) Err() error {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	if q.ctx.Err() != nil {
+		return q.ctx.Err()
+	}
+	return q.err
+}
+
+func (q *Query) insertQuery(job *bigquerypb.Job, opts []gax.CallOption) {
+	res, err := q.h.c.InsertJob(q.ctx, &bigquerypb.InsertJobRequest{
+		ProjectId: q.projectID,
+		Job:       job,
+	}, opts...)
+
+	if err != nil {
+		q.markDone(err)
+		return
+	}
+
+	q.consumeQueryResponse(&bigquerypb.GetQueryResultsResponse{
+		JobReference: res.GetJobReference(),
+	})
+
+	go q.waitForQueryBackground(opts)
+}
+
+func (q *Query) runQuery(req *bigquerypb.PostQueryRequest, opts []gax.CallOption) {
+	res, err := q.h.c.Query(q.ctx, req, opts...)
+	if err != nil {
+		q.markDone(err)
+		return
+	}
+	q.queryID = res.GetQueryId()
+
+	q.consumeQueryResponse(res)
+
+	go q.waitForQueryBackground(opts)
+}
+
+func (q *Query) waitForQueryBackground(opts []gax.CallOption) {
+	backoff := gax.Backoff{
+		Initial:    50 * time.Millisecond,
+		Multiplier: 1.3,
+		Max:        60 * time.Second,
+	}
+	for !q.complete {
+		err := q.waitForQuery(q.ctx, opts)
+		if err != nil {
+			q.markDone(err)
+			return
+		}
+		select {
+		case <-time.After(backoff.Pause()):
+		case <-q.ctx.Done():
+			q.markDone(q.ctx.Err())
+			return
+		}
+	}
+	q.markDone(nil)
+}
+
+func (q *Query) markDone(err error) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// Check if already done to prevent panic on closing closed channel.
+	select {
+	case <-q.ready:
+		// Already closed
+		return
+	default:
+		// Not closed yet
+		q.err = err
+		close(q.ready)
+	}
+}
+
+func (q *Query) waitForQuery(ctx context.Context, opts []gax.CallOption) error {
+	res, err := q.h.c.GetQueryResults(ctx, &bigquerypb.GetQueryResultsRequest{
+		ProjectId:  q.projectID,
+		JobId:      q.jobID,
+		Location:   q.location,
+		MaxResults: wrapperspb.UInt32(0),
+		FormatOptions: &bigquerypb.DataFormatOptions{
+			UseInt64Timestamp: true,
```

Comment:
We should migrate to the new field.  I'll refresh the protos to make sure its present.